### PR TITLE
Disable util::Result copying and assignment

### PR DESCRIPTION
**Pull request by ryanofsky**
This PR just contains the first two commits of #25665.

It disables copying of `util::Result` objects because unnecessary copies are inefficient and not possible after #25665, which makes `util::Result` object move-only.

It disables the assignment operator and replaces it with an `Update()` method, because #25665 adds more information to `util::Result` objects (warning and error messages and failure values) and having an assignment operator that overwrites data instead of merging it would make it easy to accidentally erase existing information while trying to assign new information.

### CODE REVIEW

**Review by stickies-v — COMMENTED** [review]
Concept ACK, breaking it out in a smaller PR seems like a good idea, thanks.

I'm unsure about removing the assignment operator altogether. I think it's good to be careful about footguns, but I'm worried about making the interface overly unintuitive.

Would it make sense to only merge another `Result` when using the `Merge` (currently called `Update`) method, and use the assignment operator to overwrite the current result? I think when people want to merge results, they'll probably want to be explicit about that, and in other cases overwriting/clearing everything seems like expected behaviour?

**Review by ryanofsky — COMMENTED** [review]
Updated 974b16b2580f4f09c8936fbead96fdbee637cb14 -> c45cb13b9e4f6eae50ab6dc42acf471921980591 ([`pr/saferesult.1`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.1) -> [`pr/saferesult.2`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/saferesult.1..pr/saferesult.2)) with suggestions

re: https://github.com/bitcoin/bitcoin/pull/29906#pullrequestreview-2011883406

> I'm unsure about removing the assignment operator altogether. I think it's good to be careful about footguns, but I'm worried about making the interface overly unintuitive.

We could add back a safer, more limited form of operator= in the future if not having it causes friction. I rewrote commit message to be clearer and mention that. But I think having a named `Update()` method should make it easier to figure out what code is doing and avoid mistakes.

> Would it make sense to only merge another `Result` when using the `Merge` (currently called `Update`) method, and use the assignment operator to overwrite the current result? I think when people want to merge results, they'll probably want to be explicit about that, and in other cases overwriting/clearing everything seems like expected behaviour?

Hmm, I'm actually not sure what this is suggesting. If there is a need to overwrite/clear results, IMO, the best way to do that would be to add a `Reset()` method. I haven't seen an actual need to do that anywhere though. The intended use for Update is to be able to write code like:

```c++
util::Result<SuccessType, FailureType> result;
if (do_something()) {
  result.Update(success_value);
} else {
  result.Update({util::Error{_("error message"), failure_value});
}
```

**Review by ryanofsky — COMMENTED** [review]
Updated c45cb13b9e4f6eae50ab6dc42acf471921980591 -> 66022315641934bda301d61f009dae9cb3b45da0 ([`pr/saferesult.2`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.2) -> [`pr/saferesult.3`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/saferesult.2..pr/saferesult.3)) adding change to actually disable the copy constructor

**Review by sedited — APPROVED** [review]
ACK 66022315641934bda301d61f009dae9cb3b45da0

**Review by furszy — COMMENTED** [review]
ACK 6602231564

**Review by stickies-v — COMMENTED** [review]
Approach ACK, code LGTM 66022315641934bda301d61f009dae9cb3b45da0 with some non-blocking suggestions.

> We could add back a safer, more limited form of operator= in the future if not having it causes friction.

I made my previous comment because I thought there are some instances of somewhat awkward `Update` usage in this PR. The majority of `util::Result` usage is/should probably not require `Update()` at all, because we're just returning either a success value or an error string. But, in those simple cases, we should just use list initialization, imo, and not assignment initialization. So, I've left a couple of code comments that remove all the instances I find awkward.

> The intended use for Update is to be able to write code like: ...

Hmmm. In the code example you give, I'd find assignment much more intuitive, since the goal doesn't seem to be combining results. If `Update()` is aimed to be used exclusively (?) when we intend to combine results, doesn't that make for a more clear interface?

**Comment by ryanofsky** [comment]
Thanks @stickies-v, I agree with your suggestions to avoid updating existing `Result` objects where possible and will apply them.

But I wonder if there are is anything else I can do to help make the interface more intuitive though. I'm not exactly clear on why it is not intuitive currently. A few places you mentioned "Update here is confusing" but I don't understand why it is confusing. Saying `result.Update(success_value)` is just meant to update the result object with a success value while keeping other result data intact. Saying `result.Update({util::Error{_("error message"), failure_value});` is meant to update the result object with an error message and failure value, again keeping other result data intact.

I think it's very similar to python dict.update function where if you start with an existing dictionary:

```python
d = {'a': 1, 'b': 2, 'c': 3}
```

and call the update method with new values:

```python
d.update({'a': 4, 'd': 5})
```

And the dictionary will be updated with the new values overwriting the old ones:

```python
{'a': 4, 'b': 2, 'c': 3, 'd': 5}
```

> In the code example you give, I'd find assignment much more intuitive, since the goal doesn't seem to be combining results. If Update() is aimed to be used exclusively (?) when we intend to combine results, doesn't that make for a more clear interface?

Update is just intended to update some data in a result object without necessarily erasing other data in the result object. Assigning is conceptually different and would imply that result being assigned to would be cleared and completely replaced by value being assigned. In this PR, the result object only has two mutually exclusive fields (success value and error string), so assigning and updating are the same thing. In #25665, result objects contain more fields so assigning and updating are different things.

File: src/init.cpp [hunk 1]
```diff
@@ -993,7 +993,7 @@ bool AppInitParameterInteraction(const ArgsManager& args)
     // ********************************************************* Step 3: parameter-to-internal-flags
     auto result = init::SetLoggingCategories(args);
     if (!result) return InitError(util::ErrorString(result));
-    result = init::SetLoggingLevel(args);
+    result.Update(init::SetLoggingLevel(args));
```
[the remark this answers is not in the document]
Comment (ryanofsky) [reply src/init.cpp]: re: https://github.com/bitcoin/bitcoin/pull/29906#discussion_r1580022711

Thanks, applied patch

File: src/test/mempool_tests.cpp [hunk 2]
```diff
@@ -260,7 +260,7 @@ BOOST_AUTO_TEST_CASE(MempoolIndexingTest)
     tx10.vout[0].scriptPubKey = CScript() << OP_11 << OP_EQUAL;
     tx10.vout[0].nValue = 10 * COIN;
 
-    ancestors_calculated = pool.CalculateMemPoolAncestors(entry.Fee(200000LL).Time(NodeSeconds{4s}).FromTx(tx10), CTxMemPool::Limits::NoLimits());
+    ancestors_calculated.Update(pool.CalculateMemPoolAncestors(entry.Fee(200000LL).Time(NodeSeconds{4s}).FromTx(tx10), CTxMemPool::Limits::NoLimits()));
```
[the remark this answers is not in the document]
Comment (ryanofsky) [reply src/test/mempool_tests.cpp]: re: https://github.com/bitcoin/bitcoin/pull/29906#discussion_r1580027359

Thanks, applied patch

File: src/wallet/test/fuzz/notifications.cpp [hunk 3]
```diff
@@ -108,9 +108,9 @@ struct FuzzedWallet {
         auto type{fuzzed_data_provider.PickValueInArray(OUTPUT_TYPES)};
         util::Result<CTxDestination> op_dest{util::Error{}};
         if (fuzzed_data_provider.ConsumeBool()) {
-            op_dest = wallet->GetNewDestination(type, "");
+            op_dest.Update(wallet->GetNewDestination(type, ""));
         } else {
-            op_dest = wallet->GetNewChangeDestination(type);
+            op_dest.Update(wallet->GetNewChangeDestination(type));
```
[the remark this answers is not in the document]
Comment (ryanofsky) [reply src/wallet/test/fuzz/notifications.cpp]: re: https://github.com/bitcoin/bitcoin/pull/29906#discussion_r1580037125

Thanks, applied patch

File: src/wallet/test/spend_tests.cpp [hunk 4]
```diff
@@ -102,7 +102,7 @@ BOOST_FIXTURE_TEST_CASE(wallet_duplicated_preset_inputs_test, TestChain100Setup)
 
     // Second case, don't use 'subtract_fee_from_outputs'.
     recipients[0].fSubtractFeeFromAmount = false;
-    res_tx = CreateTransaction(*wallet, recipients, /*change_pos=*/std::nullopt, coin_control);
+    res_tx.Update(CreateTransaction(*wallet, recipients, /*change_pos=*/std::nullopt, coin_control));
```
[the remark this answers is not in the document]
Comment (ryanofsky) [reply src/wallet/test/spend_tests.cpp]: re: https://github.com/bitcoin/bitcoin/pull/29906#discussion_r1580043133

Thanks, applied patch

**Review by ryanofsky — COMMENTED** [review]
Updated 66022315641934bda301d61f009dae9cb3b45da0 -> 4fb08c28ba0f14782fdab4ac54387aae0aba82ed ([`pr/saferesult.3`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.3) -> [`pr/saferesult.4`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/saferesult.3..pr/saferesult.4)) with suggested changes using new result objects to avoid updating existing ones where possible
Update 4fb08c28ba0f14782fdab4ac54387aae0aba82ed -> 88aa093356ff142efe60df989323eb1ed1aa9e18 ([`pr/saferesult.4`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.4) -> [`pr/saferesult.5`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/saferesult.4..pr/saferesult.5)) with no changes overall, just moving changes out of first commit that belonged in second commit

**Comment by stickies-v** [comment]
Sorry, I've been doing a poor job explaining myself. I am fully onboard with how `Update` works.  When I used the word "confusion", I was talking about _usage_, not interface*. Prior to this push, `Update()` was used in a couple of places where we know or expect the `Result` to be empty, and in those case I think we should just be explicit about that and use list initialization. When I see `result.Update()` being called, I have to look up which values are in `result` already and ensure that's correct, whereas with `auto result{...};` I know right away that we're starting from an empty result.

To stick with the python analogy: using `update` when we know or expect `a` to be an empty dictionary, is imo an anti-pattern and should be avoided.

```python
a = {}
// do some logic to calculate `value`
a.update({1: value})
```

Instead, it is much clearer to use assignment, so the reader doesn't have to figure out which values `a` held already.
```python
// do some logic to calculate `value`
a = {1: value}
```

In short: I think it would be helpful if we only use `result.Update()` when `result` (potentially) has had fields set already, and use list initialization (or if necessary assignment initialization) for when we know (and thus enforce that) `result` is empty. This is also the concern I had with the code example [at the end of this comment](https://github.com/bitcoin/bitcoin/pull/29906#pullrequestreview-2012191090).

I hope I did a better job explaining myself, but otherwise I'm happy to talk about it more offline to not pollute this PR more.

* well, except for making sure the method names align with how we intend people to use them, hence my earlier renaming suggestions.

**Comment by ryanofsky** [comment]
Rebased 88aa093356ff142efe60df989323eb1ed1aa9e18 -> 0cbcb66257db65f0a68d82d17a2d1a83d8654b38 ([`pr/saferesult.5`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.5) -> [`pr/saferesult.6`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/saferesult.5-rebase..pr/saferesult.6)) dropping Result::Update() method in this PR, and last 2 uses of it.

re: https://github.com/bitcoin/bitcoin/pull/29906#issuecomment-2078213252

Thanks for clarifying. I think we might disagree whether certain uses of an update function would be antipatterns, but I completely agree with all the code changes you suggested, and that in general if you have a choice between initializing values once or letting them change over time, you should prefer initializing them once so there's less changing state to reason about. I also looked at remaining two update calls in this PR, and thought they both were not good uses, so I removed them as well.

My [update example](https://github.com/bitcoin/bitcoin/pull/29906#pullrequestreview-2012191090]) was very simple because I wanted to show minimal code that demonstrated _how_ to use the update function, but the code was probably too simple to demonstrate _why_ you would want to use it. The use-case for update is when you have a function that is performing multiple operations and you want to accumulate warnings and errors from all the operations and return them. Examples would be the [`BlockManager::SaveBlockToDisk`](https://github.com/ryanofsky/bitcoin/blob/4d2c9de24916f8d69514ea7c7251136e2762fa5c/src/node/blockstorage.cpp#L1167-L1195) function from #29700 or the [`CompleteChainstateInitialization`](https://github.com/ryanofsky/bitcoin/blob/55d7de92bbbe035c1833c89f885af14e5b243932/src/node/chainstate.cpp#L38-L179) function from #25665. I think these are reasonable uses of the update function, but it's possible there could be better ways to write this code, and you might have another idea what it should look like.

**Review by stickies-v — APPROVED** [review]
ACK e1be443315be6ba6c80267e0e6be59deee77de50

> The use-case for update is when you have a function that is performing multiple operations and you want to accumulate warnings and errors from all the operations and return them.

I agree, and from what I've seen so far, I would prefer that to be the only use-case.

> I also looked at remaining two update calls in this PR, and thought they both were not good uses, so I removed them as well.

Ah, nice one. Given that chaining results is not actually implemented yet, it makes sense that there should be no existing usage of `Update` in these first commits. I looked at the initial implementation of these 2 new updates to make sure we're not missing anything and indeed, chaining is not relevant.

---

I've [posted](https://github.com/bitcoin/bitcoin/pull/25665#issuecomment-2079388456) my response to the second part of [your comment](https://github.com/bitcoin/bitcoin/pull/29906#issuecomment-2078452007) about `Update()` usage on #25665, since `Update()` is no longer in this PR.

File: src/validation.cpp [hunk 5]
```diff
@@ -970,11 +971,13 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
         if (ws.m_vsize > EXTRA_DESCENDANT_TX_SIZE_LIMIT) {
             return state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY, "too-long-mempool-chain", error_message);
         }
-        ancestors = m_pool.CalculateMemPoolAncestors(*entry, cpfp_carve_out_limits);
-        if (!ancestors) return state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY, "too-long-mempool-chain", error_message);
+        if (auto ancestors{m_pool.CalculateMemPoolAncestors(*entry, cpfp_carve_out_limits)}) {
```
Comment (furszy) [inline src/validation.cpp:974]: This should be shadowing the previous `ancestors` variable (why the compiler isn't complaining about this?). The else path can also access the variable when it is declared inside the if statement.

**Comment by achow101** [comment]
ACK e1be443315be6ba6c80267e0e6be59deee77de50

File: src/util/result.h [hunk 6]
```diff
@@ -39,13 +39,25 @@ class Result
...
 
+    //! Disallow copy constructor, require Result to be moved for efficiency.
+    Result(const Result&) = delete;
+
+    //! Disallow operator= to avoid confusion in the future when the Result
+    //! class gains support for richer error reporting, and callers should have
+    //! ability to set a new result value without clearing existing error
+    //! messages.
+    Result& operator=(const Result&) = default;
+    Result& operator=(Result&&) = default;
```
Comment (achow101) [inline src/util/result.h:50]: Can you briefly explain why these `operator=` are made private as opposed to deleting them?

File: src/test/mempool_tests.cpp [hunk 7]
```diff
@@ -260,9 +262,12 @@ BOOST_AUTO_TEST_CASE(MempoolIndexingTest)
...
 
-    ancestors_calculated = pool.CalculateMemPoolAncestors(entry.Fee(200000LL).Time(NodeSeconds{4s}).FromTx(tx10), CTxMemPool::Limits::NoLimits());
-    BOOST_REQUIRE(ancestors_calculated);
-    BOOST_CHECK(*ancestors_calculated == setAncestors);
+    {
+        auto ancestors_calculated{pool.CalculateMemPoolAncestors(entry.Fee(200000LL).Time(NodeSeconds{4s}).FromTx(tx10), CTxMemPool::Limits::NoLimits())};
+        BOOST_REQUIRE(ancestors_calculated);
+        BOOST_CHECK(*ancestors_calculated == setAncestors);
+    }
+
```
[the remark this answers is not in the document]
Comment (ryanofsky) [reply src/test/mempool_tests.cpp:270]: re: https://github.com/bitcoin/bitcoin/pull/29906#discussion_r1580997920

> nit: unnecessary extra newline

Thanks, removed

File: src/util/result.h (hunk 6 above)
Comment (ryanofsky) [reply src/util/result.h]: re: https://github.com/bitcoin/bitcoin/pull/29906#discussion_r1581409638

> Can you briefly explain why these `operator=` are made private as opposed to deleting them?

Changed these to `delete`. The default operators were needed in an earlier version of the PR which had an `Update()` method which called them privately.

File: src/validation.cpp (hunk 5 above)
Comment (ryanofsky) [reply src/validation.cpp:974]: re: https://github.com/bitcoin/bitcoin/pull/29906#discussion_r1581128440

> But, in any case, it's not a big deal if you don't think it's worth it. I also don't think this is a problem anyway.

I'm happy to go with any suggestion. I don't see shadowing as a problem in all cases, so this is just the best way to write the code that I could think of.

File: src/wallet/test/fuzz/notifications.cpp [hunk 8]
```diff
@@ -106,13 +106,12 @@ struct FuzzedWallet {
...
-        util::Result<CTxDestination> op_dest{util::Error{}};
         if (fuzzed_data_provider.ConsumeBool()) {
-            op_dest = wallet->GetNewDestination(type, "");
+            return *Assert(wallet->GetNewDestination(type, ""));
         } else {
-            op_dest = wallet->GetNewChangeDestination(type);
+            return *Assert(wallet->GetNewChangeDestination(type));
         }
-        return *Assert(op_dest);
+        assert(false);
```
[the remark this answers is not in the document]
Comment (ryanofsky) [reply src/wallet/test/fuzz/notifications.cpp]: re: https://github.com/bitcoin/bitcoin/pull/29906#discussion_r1581039091

> nit: Is this line needed?

Nope, dropped this

**Review by ryanofsky — COMMENTED** [review]
Updated e1be443315be6ba6c80267e0e6be59deee77de50 -> 9552619961049d7673f84c40917b0385be70b782 ([`pr/saferesult.7`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.7) -> [`pr/saferesult.8`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.8), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/saferesult.7..pr/saferesult.8)) with small suggested cleanups

**Comment by stickies-v** [comment]
Re-ACK 9552619961049d7673f84c40917b0385be70b782 , just addressing nits, no other changes

**Review by furszy — COMMENTED** [review]
ACK 9552619961049d7673f84c40917b0385be70b782

**Review by sedited — APPROVED** [review]
ACK 9552619961049d7673f84c40917b0385be70b782

File: src/validation.cpp (hunk 5 above)
Comment (ryanofsky) [reply src/validation.cpp]: re: https://github.com/bitcoin/bitcoin/pull/29906#discussion_r1581580725

> I'm happy to go with any suggestion. I don't see shadowing as a problem in all cases, so this is just the best way to write the code that I could think of.

Not sure why I couldn't think of a good way to write this without shadowing last week. But added `_retry` suffixes now so the code should be clearer and shadowing should be gone :sunglasses:

**Review by ryanofsky — COMMENTED** [review]
Updated 9552619961049d7673f84c40917b0385be70b782 -> 6a8b2befeab25e4e92d8e947a23e78014695e06c ([`pr/saferesult.8`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.8) -> [`pr/saferesult.9`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.9), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/saferesult.8..pr/saferesult.9)) tweaking variables to avoid shadowing

**Comment by stickies-v** [comment]
re-ACK 6a8b2befeab25e4e92d8e947a23e78014695e06c

**Review by furszy — COMMENTED** [review]
re-ACK https://github.com/bitcoin/bitcoin/commit/6a8b2befeab25e4e92d8e947a23e78014695e06c

**Comment by achow101** [comment]
ACK 6a8b2befeab25e4e92d8e947a23e78014695e06c
